refactor: move background detection into Lua

gpanders opened the pull request:
Now that we have the `TermResponse` event we have everything we need to do this in Lua instead of C. The Lua implementation is much more straightforward and simple.

There is a functional change: the background detection happens later in the Nvim startup process. This shouldn't matter, as in either case the `'background'` option isn't changed if it has already been set by the user.

This deprecates the `term_background` UI option. I am not removing it in this PR but I am marking it deprecated to be removed in a future release.

gpanders committed:
refactor: move background color detection into Lua

gpanders commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1176,9 +1176,85 @@ function vim._init_default_autocmds()
+    return
+  end
+
+  vim.api.nvim_create_autocmd('TermResponse', {
```
TODO: Add a timer to delete this autocommand if no response is received after some timeout

zeertzjq commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1176,9 +1176,85 @@ function vim._init_default_autocmds()
+    end,
+  })
+
+  io.stdout:write('\x1b]11;?\x1b\\')
```
This doesn't work for remote TUI.

zeertzjq commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1176,9 +1176,85 @@ function vim._init_default_autocmds()
   vim._init_default_autocmds()
+
+  for _, ui in ipairs(vim.api.nvim_list_uis()) do
+    if ui.stdout_tty then
```
```suggestion
    if ui.chan == 1 and ui.stdout_tty then
```

gpanders committed:
docs: deprecate the "term_background" UI field

clason commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1176,9 +1176,97 @@ function vim._init_default_autocmds()
+---
+--- @param c string Color as a string of hex chars
+--- @return number Intensity of the color
+local function hexcolor(c)
```
```suggestion
local function luminance(c)
```
to match the description below? (should use the same term in all docstrings and variable names)

gpanders committed:
test: use ST terminator instead of BEL in OSC sequences

libtermkey does not interpret OSC sequences that end with a BEL (0x07)
instead of an ST (0x1b 0x5c) terminator. This causes these tests to fail
since the OSC response is now parsed via libtermkey. Change the tests to
use the ST terminator to appease libtermkey.

gpanders committed:
fix(lua): do not schedule events if Nvim is exiting

If Nvim is in the process of exiting then we do not want to allocate any
new refs onto the event loop, because they will not be freed and will
result in a memory leak.

justinmk reviewed (approved):
Awesome!

justinmk commented on runtime/doc/news.txt:
```diff
@@ -340,4 +340,8 @@ release.
 • vim.treesitter.languagetree functions:
   - |LanguageTree:for_each_child()|	Use |LanguageTree:children()| (non-recursive) instead.
 
+• The "term_background" UI option |ui-ext-options| is deprecated and no longer
```
also worth mentioning in deprecated.txt

xdBronch commented:
after updating nvim recently i noticed that when running `:wq` on zig files with [zls](https://github.com/zigtools/zls) theres a large delay, 1-2 seconds, before it actually exits. `:w` and `:q` on their own dont do this and ive bisected it to 22eb2ba18336df6cd70a88f666818ee5d8ba92d2, any ideas?

zeertzjq commented:
Please provide compile flags from `:verbose version`

clason commented:
Please open a new issue and provide all the necessary information.
